### PR TITLE
qa: packaged_pins — packaged plate/boxes/pin parity check (certified≠installed instrument)

### PR DESCRIPTION
**Pull request by 100yenadmin**
## Summary
- Add the stdlib `qa/packaged_pins.py` parity instrument for packaged plate, boxes, and camera-pin data.
- Add red-first synthetic `.app`/repo coverage for the specified drift and tri-state cases.
- Document the pre-flight sync target and packaged-pins gate in the room pipeline runbook.

## Verification
- `uv run --directory servers/engine python -m pytest /Users/m1/worldos-wt-pins/qa/test_packaged_pins.py -q -p no:xdist`
  - `9 passed`
- `bash qa/fast_gate.sh`
  - `257 passed`

Logs: `/Users/m1/Codex/session-notes/2026-09-02/worldos-refresh/artifacts/pins-tool/`

## Claim class
`instrument` — proves packaged data == repo data can be asserted mechanically at the tested head SHA. It does not prove any §9 gate or that a plate renders correctly.

Auto-merge is intentionally not enabled; this is a side PR for orchestrator review.

### CODE REVIEW

**Comment by 100yenadmin** [comment]
**Cross-model review (Claude reviewing the Codex-authored diff) — ACCEPT-WITH-FIXES, fixes applied in f81ab501.**
- MERGE_BLOCKING · FIXED NOW: an I/O exception inside `check()` escaped as a traceback → exit 1 = false RED and no JSON (TOCTOU with a rebuild in progress). `main()` now maps any exception to ERROR/2 with the report written (test added).
- NON_BLOCKING · FIXED NOW: unknown or empty `--rooms` → ERROR (was a vacuous GREEN); report carries `repo` + `rooms_requested`; containment check no longer resolves symlinks (false MISSING on a linked plates/ dir); the `.app`-suffix test is red-first (real dir with a valid StreamingAssets tree); RUNBOOK-INDEX command uses an absolute script path.
- Verified live: real app + repo → `PINS GREEN (0 drift)` exit 0 (8 rooms); negative control (byte flip, ortho change, bad path, repo-only room) → `PINS RED (4 drift)` exit 1 with JSON; `--rooms typo` → ERROR/2. 12 tests pass bare. Claim class: `instrument` — parity only; proves nothing about rendering or §9 gates.

**Comment by 100yenadmin** [comment]
Correction to the count above: 11 tests pass bare-run (9 original + 2 added), not 12.

File: qa/packaged_pins.py [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+    sha = result.stdout.strip()
+    return sha if result.returncode == 0 and sha else None
+
+
+def _base_report(app: Path, repo: Path) -> dict[str, Any]:
+    return {
+        "schema_version": 1,
+        "ts": datetime.now(timezone.utc).isoformat(),
+        "app": str(app.resolve()),
+        "repo_sha": _repo_sha(repo),
```
[the remark this answers is not in the document]
Comment (100yenadmin) [reply qa/packaged_pins.py]: FIXED NOW (7c627f93): the report records repo_dirty and stamps repo_sha '-dirty' (plus a WARN) whenever extensions/renderers/unity differs from HEAD by BYTES (status entries whose raw bytes equal the committed blob are LFS clean-filter phantoms, not changes). Deliberately not a RED/ERROR: the tool also runs mid-iteration; refusing dirty trees at closeout is the citing rule (SCORECARD rows cite sha + flag).

File: qa/packaged_pins.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+    return sha if result.returncode == 0 and sha else None
+
+
+def _base_report(app: Path, repo: Path) -> dict[str, Any]:
+    return {
+        "schema_version": 1,
+        "ts": datetime.now(timezone.utc).isoformat(),
+        "app": str(app.resolve()),
+        "repo": str(repo.resolve()),
+        "repo_sha": _repo_sha(repo),
```
[the remark this answers is not in the document]
Comment (100yenadmin) [reply qa/packaged_pins.py]: FIXED NOW (7c627f93): repo_dirty flag + '-dirty' sha suffix + WARN; parity is explicitly a working-tree measurement; see the sibling thread for why it is recorded rather than refused.

File: qa/packaged_pins.py [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+    if source_entry is None:
+        reasons.append("MISSING room on source side")
+    elif packaged_entry is None:
+        reasons.append("MISSING room on packaged side")
+    elif not isinstance(source_entry, dict) or not isinstance(packaged_entry, dict):
+        reasons.append("invalid room manifest entry")
+    else:
+        _compare_file("plate", source_entry, packaged_entry, source_root, packaged_root, reasons)
+        _compare_file("boxes", source_entry, packaged_entry, source_root, packaged_root, reasons)
+        _compare_camera_pin(source_entry, packaged_entry, reasons)
```
[the remark this answers is not in the document]
Comment (100yenadmin) [reply qa/packaged_pins.py:250]: Agreed — FIXED NOW (7c627f93): every other runtime field of the room entry (planeSize, effects, door_hotspots, …) is compared; '_'-prefixed comment/provenance keys are excluded. Test added.

File: qa/packaged_pins.py [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+        packaged_manifest = _load_manifest(packaged_manifest_path)
+    except (OSError, UnicodeError, ValueError, TypeError, json.JSONDecodeError) as exc:
+        return _error(app, repo, f"packaged manifest unreadable: {exc}")
+
+    source_rooms = source_manifest["plates"]
+    packaged_rooms = packaged_manifest["plates"]
+    selected = _normalise_rooms(rooms)
+    room_names = sorted(set(source_rooms) | set(packaged_rooms))
+    if selected is not None:
+        room_names = [room for room in room_names if room in selected]
```
[the remark this answers is not in the document]
Comment (100yenadmin) [reply qa/packaged_pins.py:292]: FIXED NOW (f81ab501, before this thread rendered): unknown or empty --rooms → ERROR/2, never a vacuous GREEN. Test added.

File: docs/RUNBOOK-INDEX.md [hunk 5]
```diff
@@ -31,8 +31,9 @@
...
 | player_cert (G1 live slice) | `qa/player_cert.py --live …` against the sandbox stack (CI half in ci.yml) | free; needs a built player in the sandbox | tri-state GREEN/RED/ERROR (exit 0/1/2) + `player_cert_report.json` per assertion. The build stamp is PROVENANCE ONLY today — a missing stamp does not ERROR the run; the first #1651 slice makes it one. A harness defect is ERROR, never a verdict | — (no scores_db row is written today — #1651 / #1709 — so a GREEN here is not yet a citable G1 run) | charter #1651; OPERATIONS install gate; PRODUCT-ROADMAP §9 G1 |
+| packaged pins | `uv run --directory servers/engine python qa/packaged_pins.py <WorldOSPlayer.app>` | free; before every sandbox gate, install, and closeout | JSON report with per-room parity and GREEN/RED/ERROR verdict | — (gate; no scores row) | ROOM-PIPELINE-RUNBOOK step 10; #1651 |
```
[the remark this answers is not in the document]
Comment (100yenadmin) [reply docs/RUNBOOK-INDEX.md]: FIXED NOW (f81ab501): the documented command uses an absolute script path ($(git rev-parse --show-toplevel)/qa/packaged_pins.py) — --directory moves the cwd, as you say.

File: qa/packaged_pins.py [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+    packaged_path = _entry_path(packaged_root, packaged_value)
+    if source_value != packaged_value:
+        reasons.append(f"{kind} path differs: source={source_value} packaged={packaged_value}")
+    if source_path is None or not source_path.is_file():
+        reasons.append(f"MISSING {kind} on source: {source_value}")
+    if packaged_path is None or not packaged_path.is_file():
+        reasons.append(f"MISSING {kind} on packaged side: {packaged_value}")
+    if source_value == packaged_value and source_path is not None and packaged_path is not None:
+        if source_path.is_file() and packaged_path.is_file():
+            if _sha256(source_path) != _sha256(packaged_path):
```
[the remark this answers is not in the document]
Comment (100yenadmin) [reply qa/packaged_pins.py:197]: FIXED NOW (f81ab501): any exception inside check() maps to ERROR/2 with the JSON report still written; a hashing failure is never a RED. Test added.

File: qa/packaged_pins.py [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+
+def _entry_path(root: Path, value: Any) -> Path | None:
+    if not isinstance(value, str) or not value:
+        return None
+    candidate = root / value
+    norm = os.path.normpath(str(candidate))
+    root_norm = os.path.normpath(str(root))
+    if norm != root_norm and not norm.startswith(root_norm + os.sep):
+        return None  # escapes the packaged/repo root
+    return candidate
```
[the remark this answers is not in the document]
Comment (100yenadmin) [reply qa/packaged_pins.py:144]: FIXED NOW (7c627f93): a device/socket target is refused before any read (is_file() is false → the plate is MISSING → RED; regression test symlinks a plate to /dev/zero and asserts no read, exit 1), oversized regular files are refused (ERROR), and '..' escapes are rejected by normalized containment. Symlinked plates/ dirs on linked volumes stay supported on purpose — the other reviewer showed resolve()-containment reports them as MISSING; both roots are our own build/repo data.

File: qa/packaged_pins.py [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+    if app.suffix != ".app":
+        return _error(app, repo, "app path is not a .app")
+
+    packaged_root = app / _STREAMING_RELATIVE
+    if not packaged_root.is_dir():
+        return _error(app, repo, "StreamingAssets directory is missing")
+
+    source_root = repo / _UNITY_RELATIVE
+    source_manifest_path = source_root / "plates_manifest.json"
+    packaged_manifest_path = packaged_root / "plates_manifest.json"
```
[the remark this answers is not in the document]
Comment (100yenadmin) [reply qa/packaged_pins.py:274]: FIXED NOW (469bfc0a) for effects_registry.json: compared by sha256 as a synthetic '__effects_registry.json' row (RED on drift or one-sided presence). A stale stage.json left in StreamingAssets is a separate EnsurePackaged clean-up question (it never deletes previously staged files) — NON_BLOCKING · ACCEPTED FOLLOW-UP, noted on #1651 with the other packaging assertions.

File: qa/packaged_pins.py [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+
+
+def _base_report(app: Path, repo: Path) -> dict[str, Any]:
+    return {
+        "schema_version": 1,
+        "ts": datetime.now(timezone.utc).isoformat(),
+        "app": str(app.resolve()),
+        "repo": str(repo.resolve()),
+        "repo_sha": (lambda sha, dirty: (sha + "-dirty") if (sha and dirty) else sha)(_repo_sha(repo), _repo_dirty(repo)),
+        "repo_dirty": _repo_dirty(repo),
```
[the remark this answers is not in the document]
Comment (100yenadmin) [reply qa/packaged_pins.py]: FIXED NOW (469bfc0a): dirtiness is computed once and cached; when it cannot be determined the sha is stamped '-unverified' (never a bare HEAD sha) and repo_dirty is null in the report. Kept as a stamp rather than an ERROR because parity is a working-tree measurement — the citing rule refuses unverified/dirty stamps at closeout.

File: qa/packaged_pins.py [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+    source_entry: dict[str, Any],
+    packaged_entry: dict[str, Any],
+    source_root: Path,
+    packaged_root: Path,
+    reasons: list[str],
+) -> None:
+    source_value = source_entry.get(kind, _MISSING)
+    packaged_value = packaged_entry.get(kind, _MISSING)
+    if source_value is _MISSING and packaged_value is _MISSING:
+        return
```
[the remark this answers is not in the document]
Comment (100yenadmin) [reply qa/packaged_pins.py:173]: Agreed — FIXED NOW (469bfc0a): a room entry without a plate on either side is RED ('no plate entry (room would render without a backdrop)'). Test added.

File: docs/ROOM-PIPELINE-RUNBOOK.md [hunk 11]
```diff
@@ -423,6 +423,7 @@ PNGs live under `extensions/renderers/unity/plates/`.
 and adding a `"stage"` key to a plate entry above is a silent no-op.
 
 **★ BOX BUILD PRE-FLIGHT (measured gap — stale manifests shipped in Box Cycles 2 AND 3, 2026-07-15):**
+- The pre-flight sync target is the Unity project ROOT (the `EnsurePackaged` source), never `Assets/StreamingAssets`; verify with `packaged_pins` after every build.
```
[the remark this answers is not in the document]
Comment (100yenadmin) [reply docs/ROOM-PIPELINE-RUNBOOK.md:426]: FIXED NOW (469bfc0a): the pre-flight sync list now names boxes/*.json and registry.json explicitly (and records that the omission is exactly how the 2026-09-02 build shipped the pre-kit sidecars).

File: qa/packaged_pins.py [hunk 12]
```diff
@@ -0,0 +1,364 @@
...
+
+
+def _entry_path(root: Path, value: Any) -> Path | None:
+    if not isinstance(value, str) or not value:
+        return None
+    candidate = root / value
+    norm = os.path.normpath(str(candidate))
+    root_norm = os.path.normpath(str(root))
+    if norm != root_norm and not norm.startswith(root_norm + os.sep):
+        return None  # escapes the packaged/repo root
```
[the remark this answers is not in the document]
Comment (100yenadmin) [reply qa/packaged_pins.py:143]: Valid edge (posted after merge): a symlink to a regular file OUTSIDE the bundle passes lexical containment and hashes the target. NON_BLOCKING · ACCEPTED FOLLOW-UP on #1651: containment will use the realpath of BOTH the candidate and the packaged root, allowing a symlinked plates/ DIRECTORY only when its realpath stays inside the bundle's realpath, and reporting 'symlink escapes bundle' as RED otherwise. Our hot-load recipe copies files, it never symlinks them, so no current evidence is affected.

File: docs/RUNBOOK-INDEX.md [hunk 13]
```diff
@@ -31,8 +31,9 @@
...
 | player_cert (G1 live slice) | `qa/player_cert.py --live …` against the sandbox stack (CI half in ci.yml) | free; needs a built player in the sandbox | tri-state GREEN/RED/ERROR (exit 0/1/2) + `player_cert_report.json` per assertion. The build stamp is PROVENANCE ONLY today — a missing stamp does not ERROR the run; the first #1651 slice makes it one. A harness defect is ERROR, never a verdict | — (no scores_db row is written today — #1651 / #1709 — so a GREEN here is not yet a citable G1 run) | charter #1651; OPERATIONS install gate; PRODUCT-ROADMAP §9 G1 |
+| packaged pins | `uv run --directory servers/engine python "$(git rev-parse --show-toplevel)/qa/packaged_pins.py" <WorldOSPlayer.app>` (ABSOLUTE script path — the runner's cwd is servers/engine) | free; before every sandbox gate, install, and closeout | JSON report with per-room parity and GREEN/RED/ERROR verdict | — (gate; no scores row) | ROOM-PIPELINE-RUNBOOK step 10; #1651 |
```
[the remark this answers is not in the document]
Comment (100yenadmin) [reply docs/RUNBOOK-INDEX.md:36]: Agreed (posted after merge): the registered command should carry --json so the evidence artifact the row promises actually exists. Carried as a one-line doc follow-up with the symlink fix (tracked on #1651).
